PyGA/genetic_algorithm.py

```diff
@@ -0,0 +1,596 @@
+import sys
+import copy
+import pickle
+import random
+import csv
+
+from pathlib import Path
+
+import yaml
+import numpy as np
+from pathlib import Path
+
+from horn_antenna import HornAntenna
+#import bicone_antenna
+#import hpol_antenna
+
+class GA:
+    def __init__(self, run_name, settingsfile='configs/settings.yaml', 
+                 initialization=None, gen=None):
+        self.run_name = run_name
+        self.custom_init = initialization
+        self.initialize_settings(settingsfile)
+        self.run_dir = Path(self.settings["run_dir"])
+        
+        if gen is not None:
+            self.generation = gen
+        else:
+            self.generation = 0
+        
+        self.population = []
+        self.best_individual = None
+        self.best_fitness = 0.0
+        
+        if self.settings["test_loop"]:
+            self.comparison = np.array([])
+            self.load_compairson()
+            self.make_run_directory()
+            self.initialize_tracker()
+    
+    
+    ### Initialization ########################################################
+    
+    def initialize_settings(self, settingsfile):
+        '''Initialize the settings from the settings file.'''
+        settingspath = Path(settingsfile)
+        if settingspath.exists():
+            with open(settingspath, 'r') as file:
+                settings = yaml.load(file, Loader=yaml.FullLoader)
+        else:
+            sys.exit('Settings file not found. Exiting.')
+            
+        if self.check_settings(settings):
+            self.settings = settings
+        else:
+            sys.exit('Invalid settings. Exiting.')
+    
+    
+    def check_settings(self, settings):
+        '''Check if the settings are valid.'''
+        if (settings['crossover_rate'] + settings['mutation_rate'] +
+            settings['reproduction_rate']) > 1.0:
+            print('Operator rates exceed 1.0.')
+            return False
+        if (settings['tournament_rate'] + settings['roulette_rate'] + 
+            settings['rank_rate']) != 1.0:
+            print('Selection rates do not sum to 1.0.')
+            return False
+        if settings['steady_state'] and settings['reproduction_rate'] != 0:
+            print('Steady state and reproduction not compatible.')
+            return False
+        return True
+    
+    
+    def initialize_population(self):
+        '''Initialize the population of horn antennas.'''
+        antenna_type = self.settings['a_type']
+        if self.custom_init is None:
+            for i in range(self.settings["npop"]):
+                antenna = self.make_antenna(antenna_type)
+                antenna.initialize()
+                self.population.append(antenna)
+        else:
+            # load in the population from a file
+            init_path = Path(f"initializations/{self.custom_init}.pkl")
+            self.load_population(init_path)
+    
+    
+    def load_compairson(self):
```

Comment:
Think you wanted this to be load_comparison, although it is called correctly on line 36